Add "Use Thumbor" blog post (#1490)
GoogleChrome/web.dev: src/site/content/en/blog/use-thumbor/index.md

2. Compress the image using Thumbor's default compression settings: <a href="http://34.67.235.246:8888/unsafe/filters:quality()/https://web.dev/backdrop-filter/hero.jpg" target="_blank" rel="noreferrer">http://34.67.235.246:8888/unsafe/filters:quality()/https://web.dev/backdrop-filter/hero.jpg</a>

<figure class="w-figure  w-figure--center">
  <img src="./Use-thumbor6.jpg" alt="Compressed image with no noticible quality issues" class="w-screenshot">
  <figcaption>Compressed image</figcaption>
</figure>

### Resizing

To resize an image while maintaining its original proportions use the format `$WIDTHx0` or `0x$HEIGHT` within the `size` portion of the URL string.

Try it now:

1. Resize the image to a width of 200 pixels while maintaining original proportions: <a href="http://34.67.235.246:8888/unsafe/200x0/https://web.dev/backdrop-filter/hero.jpg" target="_blank" rel="noreferrer">http://34.67.235.246:8888/unsafe/200x0/https://web.dev/backdrop-filter/hero.jpg</a>

<figure class="w-figure  w-figure--center">
  <img src="./Use-thumbor7.jpg" alt="Image that is 200 pixels wide" class="w-screenshot">
  <figcaption>Image resized to a width of 200 pixels</figcaption>
</figure>

> 2. Resize the image to a height of 500 pixels while maintaining original proportion<a href="http://34.67.235.246:8888/unsafe/0x500/https://web.dev/backdrop-filter/hero.jpg" target="_blank" rel="noreferrer">http://34.67.235.246:8888/unsafe/0x500/https://web.dev/backdrop-filter/hero.jpg</a>

Reviewer: ```suggestion
2. Resize the image to a height of 500 pixels while maintaining original proportion: <a href="http://34.67.235.246:8888/unsafe/0x500/https://web.dev/backdrop-filter/hero.jpg" target="_blank" rel="noreferrer">http://34.67.235.246:8888/unsafe/0x500/https://web.dev/backdrop-filter/hero.jpg</a>
```

Just missing a colon and a space

  <figcaption>Image resized to a width of 100 pixels</figcaption>
</figure>

These methods are just a few of Thumbor's many cropping and resizing options. To read about other options, check out [Usage](https://github.com/thumbor/thumbor/wiki/Usage).

### File Formats

The [format](https://thumbor.readthedocs.io/en/latest/format.html) filter converts images to `jpeg`, `webp`, `gif`, or `png`. Keep in mind that if you're optimizing for performance you should [use either JPEG or WebP](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) as PNG and GIF files tend to be significantly larger and do not compress as well.

Try it now:

1. Convert the image to WebP. If you open the **Network** panel of DevTools the document's **Content-Type response header** shows that the server returned a WebP image: <a href="http://34.67.235.246:8888/unsafe/filters:format(webp)/https://web.dev/backdrop-filter/hero.jpg" target="_blank" rel="noreferrer">http://34.67.235.246:8888/unsafe/filters:format(webp)/https://web.dev/backdrop-filter/hero.jpg</a>

<figure class="w-figure  w-figure--center">

>   <img src="./Use-thumbor11.jpg" alt="DevTools screenshot showing the content-type (WebP) of an image" class="w-screenshot">

Reviewer: This image looks low-quality but I can fix it in a subsequent PR

cc @robdodson

Author: I updated this image. It's still not super high quality bc I was trying to keep the file size reasonable but it's definitely better than the previous version.

>   <figcaption>The `content-type` header in DevTools</figcaption>

Reviewer: ```suggestion
  <figcaption>The <code>content-type</code> header in DevTools</figcaption>
```

Markdown doesn't work inside of HTML

</figure>

## Next Steps

Try out other [filters](https://thumbor.readthedocs.io/en/latest/filters.html) and transformations on the `hero.jpg` image.

If you're following along using your own Thumbor installation, check out the appendix below that explains how and why to use the `thumbor.conf` file.

## Appendix: `thumbor.conf`

Many of the configuration options discussed in this post, plus many others, can be established as defaults by setting up and using a `thumbor.conf` configuration file. Settings in the `thumbor.conf` file will be applied to all images unless overridden by the URL string parameters.

1. Run the `thumbor-config` command to create a new `thumbor.conf` file.

```bash
thumbor-config > ./thumbor.conf
```

2. Open your new `thumbor.conf` file. The `thumbor-config` command generated a file that lists and explains all Thumbor configuration options.

3. Configure settings by uncommenting lines and changing the default values.
You may find it useful to set the following settings:
* `QUALITY`
* `AUTO_WEBP`
* `MAX_WIDTH` and `MAX_HEIGHT`

> * `ALLOW_ANIMATED_GIFS`

Reviewer: These bullet-list items are formatting weirdly but I believe I was the one that suggested changing it to a bulleted list so I'll also fix this in a subsequent PR

cc @robdodson

Author: I figured out the fix for this so need for a separate PR.

---
title: Optimize images with Thumbor
subhead: Thumbor can be used for free to resize, compress, and transform images on-demand.
authors:
  - katiehempenius
date: 2019-09-20

> hero: hero.jpg

Reviewer 2: I think this is the same hero image that was used in the other thumbor post. What do you think about swapping it?

Author: Updated ✔️

## Prequisites

This post assumes that you understand how image CDNs can improve your load performance. If not, check out [Use image CDNs to optimize images](https://web.dev/image-cdns). It also assumes that you've built basic websites before.

{% Aside %}

If you would like to install Thumbor on your own server and then follow along with this post, check out [How to install the Thumbor image CDN](https://web.dev/install-thumbor). Whenever you see `http://34.67.235.246:8888` in this post you'll need to replace that origin with your Thumbor instance's origin.

{% endAside %}

## Thumbor URL Format

As mentioned in [Use Image CDNs to Optimize Images](https://web.dev/image-cdns), each image CDN uses a slightly different URL format for images. Figure 1 represents Thumbor's format.

<figure class="w-figure  w-figure--center">
  <img src="./Use-thumbor0.jpg" alt="A Thumbor URL has the following components: origin, security key, size, filters and image." class="w-screenshot">
</figure>

> _Figure 1. Thumbor's URL format_

Reviewer 2: one small note: I think we stopped doing the `Figure 1.` style of figcaptions a while back. I think you can omit that and just use "Thumbor's URL format".

> _Figure 1. Thumbor's URL format_

Reviewer 2: You can look at the image docs for examples. https://web.dev/handbook/markup-media these no longer use the `Figure X` style.

> _Figure 1. Thumbor's URL format_

Reviewer: Thanks, Rob. I thought I had seen it on other pages but `grep` tells me that I'm mistaken. Just a habit from developers.google.com/web.

### Security Key

The `unsafe` part of the URL indicates that you're using Thumbor without a [security key](https://github.com/thumbor/thumbor/wiki/security). A security key prevents a user from making unauthorized changes to your image URLs. By changing the image URL, a user could use your server (and your hosting bill) to resize their images, or, more maliciously, to overload your server. This guide won't cover setting up Thumbor's security key feature.

### Size

This part of the URL specifies the desired size of the output image. This can be omitted if you don't want to change the size of the image. Thumbor will use different approaches like cropping or scaling to achieve the desired size depending on the other URL parameters. The next section of this post explains how to resize images in more detail.

Try it now:

> 1. Click the URL below to view the image served at its original size. The image opens in a new tab.

Reviewer: ```suggestion
1. Click the following URL to view the image served at its original size in a new tab: <a href="http://34.67.235.246:8888/unsafe/https://web.dev/backdrop-filter/hero.jpg" target="_blank" rel="noreferrer">http://34.67.235.246:8888/unsafe/https://web.dev/backdrop-filter/hero.jpg</a>
```

Now that I see the rendered article, I realize it's a little awkward to have the URL below the instructions. Please use the format above for all the instructions. So each instruction should end with a colon, and then we list the URL to be clicked

<a href="http://34.67.235.246:8888/unsafe/https://web.dev/backdrop-filter/hero.jpg" target="_blank" rel="noreferrer">http://34.67.235.246:8888/unsafe/https://web.dev/backdrop-filter/hero.jpg</a>

<figure class="w-figure  w-figure--center">

>   <img src="./Use-thumbor1.jpg" alt="Image at original size" class="w-screenshot">

Reviewer: Please add a `figcaption` for all images in the post

  <img src="./Use-thumbor10.jpg" alt="Image that is 100 pixels wide" class="w-screenshot">
</figure>

These methods are just a few of Thumbor's many cropping and resizing options. To read about other options, check out [Usage](https://github.com/thumbor/thumbor/wiki/Usage).

### File Formats

The [format](https://thumbor.readthedocs.io/en/latest/format.html) filter converts images to `jpeg`, `webp`, `gif`, or `png`. Keep in mind that if you're optimizing for performance you should [use either JPEG or WebP](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) as PNG and GIF files tend to be significantly larger and do not compress as well.

Try it now:

1. Convert the image to WebP. If you open the **Network** panel of DevTools the document's **Content-Type response header** shows that the server returned a WebP image.

<a href="http://34.67.235.246:8888/unsafe/filters:format(webp)/https://web.dev/backdrop-filter/hero.jpg" target="_blank" rel="noreferrer">http://34.67.235.246:8888/unsafe/filters:format(webp)/https://web.dev/backdrop-filter/hero.jpg</a>

<figure class="w-figure  w-figure--center">

>   <img src="./Use-thumbor11.jpg" alt="DevTools screenshot showing the content-type (WebP) of an image" class="w-screenshot">

Reviewer: Can you use this screenshot instead? The `content-type` text is hard to read. You'll need to add the red box around `content-type` again.

![image](https://user-images.githubusercontent.com/4713486/64937762-1a514480-d88e-11e9-850e-323ceb3ec411.png)

</figure>

## Next Steps

Try out other [filters](https://thumbor.readthedocs.io/en/latest/filters.html) and transformations on the `hero.jpg` image.

> If you're following along using your own Thumbor installation, check out the appendix below that explains how and why to use the thumbor.conf file.

Reviewer: ```suggestion
If you're following along using your own Thumbor installation, check out the appendix below that explains how and why to use the `thumbor.conf` file.
```

> ## Appendix: Thumbor.conf

Reviewer: ```suggestion
## Appendix: `thumbor.conf`
```

> Many of the configuration options discussed in this post, plus many others, can be established as defaults by setting up and using the `thumbor.conf` file. Settings in the `thumbor.conf` file will be applied to all images unless overridden by the URL string parameters.

Reviewer: ```suggestion
Many of the configuration options discussed in this post, plus many others, can be established as defaults by setting up and using a `thumbor.conf` configuration file. Settings in the `thumbor.conf` file will be applied to all images unless overridden by the URL string parameters.
```

> The best way to start setting up your `thumbor.conf` file is to use the `thumbor-config` command to create a new config.

Reviewer: ```suggestion
1. Run the `thumbor-config` command to create a new `thumbor.conf` file.
```

```bash
thumbor-config > ./thumbor.conf
```

> The comments in this config will list and explain all Thumbor configurations options and their defaults. To change a setting, uncomment the line and change the text. The following are a few of the settings that you may find useful to set: `QUALITY`, `AUTO_WEBP`, `MAX_WIDTH` and `MAX_HEIGHT`, and `ALLOW_ANIMATED_GIFS`.

Reviewer: ```suggestion
1. Open your new `thumbor.conf` file. The `thumbor-config` command listed and explained all Thumbor configuration options.
1. Configure settings by uncommenting lines and changing the default values.

You may find it useful to set the following settings:

* `QUALITY`
* `AUTO_WEBP`
* `MAX_WIDTH` and `MAX_HEIGHT`
* `ALLOW_ANIMATED_GIFS`
```

> In order for the thumbor.conf settings to apply, Thumbor must be run using the `--conf` flag. This tells Thumbor the location of the thumbor.conf file. 

Reviewer: ```suggestion
1. Run Thumbor with the `--conf` flag in order to apply your `thumbor.conf` settings.
```
